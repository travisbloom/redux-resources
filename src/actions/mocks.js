import {
    mockStore,
    normalizer,
    RESOURCE_REDUCER,
    RESOURCE,
    RESOURCE_ID,
    INITIAL_STATE,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    QUERY,
    RESOURCE_UPDATE_PAYLOAD,
    formatErrors,
} from '../mocks'
import createResource from './createResource'
import deleteResource from './deleteResource'
import listResource from './listResource'
import partialUpdateResource from './partialUpdateResource'
import updateResource from './updateResource'
import retrieveResource from './retrieveResource'

const generateCreateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState)
    const { createMockResources } = createResource({
        resource: RESOURCE_REDUCER,
        normalizer,
        request,
        formatErrors,
    })
    await store.dispatch(createMockResources(RESOURCE, options))
    return store.getActions()
}

const generateDeleteResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { deleteMockResources } = deleteResource({
        resource: RESOURCE_REDUCER,
        request,
        formatErrors,
    })
    await store.dispatch(deleteMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generateListResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { listMockResources } = listResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(listMockResources(QUERY, options))
    return store.getActions()
}

const generateRetrieveResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { retrieveMockResources } = retrieveResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(retrieveMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generatePartialUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { partialUpdateMockResources } = partialUpdateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(partialUpdateMockResources(RESOURCE_ID, RESOURCE_UPDATE_PAYLOAD, options))
    return store.getActions()
}

const generateUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { updateMockResources } = updateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    const body = { ...RESOURCE, ...RESOURCE_UPDATE_PAYLOAD }
    await store.dispatch(updateMockResources(RESOURCE_ID, body, options))
    return store.getActions()
}

module.exports = {
    generateCreateResourceActions,
    generateDeleteResourceActions,
    generateListResourceActions,
    generateRetrieveResourceActions,
    generatePartialUpdateResourceActions,
    generateUpdateResourceActions,
}
