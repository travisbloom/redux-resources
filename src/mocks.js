import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Schema, normalize, arrayOf } from 'normalizr'

const RESOURCE_REDUCER = 'resources'

const RESOURCE_ID = 'resourceId'
const RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
}

const RESOURCE_UPDATE_PAYLOAD = {
    id: RESOURCE_ID,
    prop1: 'updated prop1',
    prop2: 'updated prop1',
}

const SERVER_ERRORS = [
    'This is an initial error',
    'This is another error',
]

const resourceSchema = new Schema(RESOURCE_REDUCER)
const normalizer = (resources) => {
    const array = Array.isArray(resources) ? resources : [resources]
    return normalize(array, arrayOf(resourceSchema))
}
const formatErrors = errors => errors

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

module.exports = {
    RESOURCE_REDUCER,
    RESOURCE_ID,
    RESOURCE,

    RESOURCE_UPDATE_PAYLOAD,

    SERVER_ERRORS,

    mockStore,
    formatErrors,
    normalizer,
}
