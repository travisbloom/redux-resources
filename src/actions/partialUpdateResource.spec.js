import {
    RESOURCE,
    SERVER_ERRORS,
    RESOURCE_UPDATE_PAYLOAD,
} from '../mocks'
import { generatePartialUpdateResourceActions } from './mocks'

const request = () => Promise.resolve({ ...RESOURCE, ...RESOURCE_UPDATE_PAYLOAD })
const errorRequest = () => Promise.reject(SERVER_ERRORS)

it('will dispatch two actions on success', async () => {
    const actions = await generatePartialUpdateResourceActions({ request })
    expect(actions.length).toBe(2)
})

it('will dispatch one action on success if its an optimistic update', async () => {
    const actions = await generatePartialUpdateResourceActions({
        request,
        isOptimisticUpdate: true,
    })
    expect(actions.length).toBe(1)
})

it('will dispatch two actions on error', async () => {
    const actions = await generatePartialUpdateResourceActions({ request: errorRequest })
    expect(actions.length).toBe(2)
})

it('will have an initial action', async () => {
    const [inititalAction] = await generatePartialUpdateResourceActions({ request })
    expect(inititalAction).toMatchSnapshot()
})

it('will have a success action', async () => {
    const [_, successAction] = await generatePartialUpdateResourceActions({ request })
    expect(successAction).toMatchSnapshot()
})

it('will have a error action', async () => {
    const [_, errorAction] = await generatePartialUpdateResourceActions({ request: errorRequest })
    expect(errorAction).toMatchSnapshot()
})

it('will have a error action with previous info if its an optimistic update', async () => {
    const [_, errorAction] = await generatePartialUpdateResourceActions({
        request: errorRequest,
        isOptimisticUpdate: true,
    })
    expect(errorAction).toMatchSnapshot()
})
