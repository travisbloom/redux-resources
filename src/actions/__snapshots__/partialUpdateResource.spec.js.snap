exports[`test will have a error action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "reduxResourcesActionType": "PARTIAL_UPDATE_RESOURCE_REQUEST_ERROR",
    "resource": "mockResources",
    "timestamp": "2016-01-01T00:00:00.000Z",
  },
  "payload": Array [
    "This is an initial error",
    "This is another error",
  ],
  "type": "PARTIAL_UPDATE_MOCK_RESOURCES_REQUEST_ERROR",
}
`;

exports[`test will have a error action with previous info if its an optimistic update 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "isOptimisticUpdate": true,
    "reduxResourcesActionType": "PARTIAL_UPDATE_RESOURCE_REQUEST_ERROR",
    "replacedLastUpdatedAt": null,
    "replacedResource": Object {
      "id": "resourceId",
      "prop1": "initial prop1",
      "prop2": "initial prop2",
      "relatedResource": "relatedResourceId",
    },
    "resource": "mockResources",
    "timestamp": "2016-01-01T00:00:00.000Z",
  },
  "payload": Array [
    "This is an initial error",
    "This is another error",
  ],
  "type": "PARTIAL_UPDATE_MOCK_RESOURCES_REQUEST_ERROR",
}
`;

exports[`test will have a success action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "normalizedResponse": Object {
      "entities": Object {
        "mockResources": Object {
          "resourceId": Object {
            "id": "resourceId",
            "prop1": "initial prop1",
            "prop2": "UPDATED PROP_1",
            "relatedResource": "relatedResourceId",
          },
        },
        "relatedMockResources": Object {
          "relatedResourceId": Object {
            "id": "relatedResourceId",
            "relatedProp1": "initial relatedProp1",
          },
        },
      },
      "result": Array [
        "resourceId",
      ],
    },
    "reduxResourcesActionType": "PARTIAL_UPDATE_RESOURCE_REQUEST_SUCCESS",
    "resource": "mockResources",
    "timestamp": "2016-01-01T00:00:00.000Z",
  },
  "payload": Object {
    "id": "resourceId",
    "prop1": "initial prop1",
    "prop2": "UPDATED PROP_1",
    "relatedResource": Object {
      "id": "relatedResourceId",
      "relatedProp1": "initial relatedProp1",
    },
  },
  "type": "PARTIAL_UPDATE_MOCK_RESOURCES_REQUEST_SUCCESS",
}
`;

exports[`test will have an initial action 1`] = `
Object {
  "meta": Object {
    "id": "resourceId",
    "reduxResourcesActionType": "PARTIAL_UPDATE_RESOURCE_REQUEST",
    "resource": "mockResources",
    "timestamp": "2016-01-01T00:00:00.000Z",
  },
  "payload": "resourceId",
  "type": "PARTIAL_UPDATE_MOCK_RESOURCES_REQUEST",
}
`;
